heron/schedulers/src/java/com/twitter/heron/scheduler/yarn/HeronMasterDriver.java

```diff
@@ -318,10 +324,79 @@ void submitHeronExecutorTask(ActiveContext context) {
         .set(HeronTaskConfiguration.CLUSTER, cluster)
         .set(HeronTaskConfiguration.PACKED_PLAN, getPackingAsString())
         .set(HeronTaskConfiguration.COMPONENT_RAM_MAP, getComponentRamMap())
-        .set(HeronTaskConfiguration.CONTAINER_ID, id)
+        .set(HeronTaskConfiguration.CONTAINER_ID, workerId)
         .set(HeronTaskConfiguration.VERBOSE, verboseMode)
         .build();
-    context.submitTask(taskConf);
+    worker.context.submitTask(taskConf);
+  }
+
+  /**
+   * {@link HeronWorker} is a data class which connects reef ids, heron ids and related objects.
+   * All the pointers in an instance are related to one container. A container is a reef object,
+   * owns one heron worker id and its handlers
+   */
+  private static final class HeronWorker {
+    private String workerId;
+    private int cores;
+    private int mem;
+    private AllocatedEvaluator evaluator;
+    private ActiveContext context;
+
+    HeronWorker(String id, int cores, int mem) {
+      this.workerId = id;
+      this.cores = cores;
+      this.mem = mem;
+    }
+  }
+
+  /**
+   * {@link MultiKeyWorkerMap} is a helper class to provide multi key lookup of a
+   * {@link HeronWorker} instance. It also ensures thread safety and update order.
+   */
+  private static final class MultiKeyWorkerMap {
+    private Map<String, HeronWorker> workerMap = new HashMap<>();
+    private Map<String, HeronWorker> evaluatorWorkerMap = new HashMap<>();
+
+    void assignEvaluatorToWorker(HeronWorker worker, AllocatedEvaluator evaluator) {
+      worker.evaluator = evaluator;
+      synchronized (workerMap) {
+        workerMap.put(worker.workerId, worker);
+        evaluatorWorkerMap.put(evaluator.getId(), worker);
+      }
+    }
+
+    HeronWorker lookupByEvaluatorId(String evaluatorId) {
+      synchronized (workerMap) {
+        return evaluatorWorkerMap.get(evaluatorId);
+      }
+    }
+
+    HeronWorker lookupByWorkerId(String workerId) {
+      HeronWorker worker;
+      synchronized (workerMap) {
+        worker = workerMap.get(workerId);
+      }
+      if (worker == null) {
+        LOG.log(Level.WARNING, "Container for executor id: {0} not found.", workerId);
```

Comment:
If this is a condition that should not ever happen, that's more of a reason IMO to throw an verbose exception than to log and return null and make all other callers above have to check for null. The cleanest implementation in my opinion is to annotate methods with @NotNull and they either do what's expected or they throw. The calling code never needs to check for null and NPEs won't ever occur.